src/vm/gcenv.ee.h

```diff
@@ -62,6 +62,10 @@ class GCToEEInterface : public IGCToCLR {
     bool CreateThread(void (*threadStart)(void*), void* arg, bool is_suspendable, const char* name);
     void WalkAsyncPinnedForPromotion(Object* object, ScanContext* sc, promote_func* callback);
     void WalkAsyncPinned(Object* object, void* context, void(*callback)(Object*, Object*, void*));
+    void FireGcStartAndGenerationRanges(uint32_t count, uint32_t depth, uint32_t reason, uint32_t type);
```

Comment:
I think to the extent possible, it would be good to compute the actual event you want to generate inside of the GC instead of having it done in the EE.  Then you can control the behavior of the events inside of the GC instead of having to make changes to the EE.  It also cleans up the GCtoEE interface because you only need generic eventing mechanisms instead of having specific event functions.  It doesn't actually look like these events are doing much interesting on the EE side.

Reply:
Absolutely agreed - I've run into a number of problems when trying to do this, though:

1. GcStart - A "client sequence number" is acquired from a [VM global](https://github.com/dotnet/coreclr/blob/master/src/vm/eventtrace.cpp#L894-L902) that is set whenever a GC is induced via the GCHeapCollect keyword. In order to calculate GcStart's payload entirely on the GC side, we'll need to ask the EE for the correct sequence number to use.
2. `GetClrInstanceId` - Almost every event sends this in event payloads. There will need to be some way for the GC to query the EE for this.
3. `PinPlugAtGCTime` and `AllocationTick` - Both of these get the type handle of the most recently allocated object and then extract the name of that type.

I'm still trying to work this out but it's not looking very pretty. At a minimum it looks like `GetClrInstanceId` will need to be on `IGCToCLR` since almost every event uses it.

Reply:
Yeah, makes sense.